rust/gitxetcore/src/upgrade_checks.rs

```diff
@@ -0,0 +1,431 @@
+
+            if version_is_newer(&version_tag, local_version) {
+                info!("get_newer_client_release_versions: Release {version_tag} newer than {local_version}.");
+                debug!(
+                    "get_newer_client_release_versions: Version {version_tag}; Body = {:?}",
+                    &release.body
+                );
+
+                newer_releases.push((version_tag, release.body));
+            } else {
+                break;
+            }
+        }
+
+        Some(newer_releases)
+    } else {
+        None
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VersionCheckInfo {
+    latest_version: String,
+    contains_critical_fix: bool,
+    query_time: DateTime<Utc>,
+    inform_time: DateTime<Utc>,
+
+    #[serde(skip)]
+    version_check_filename: PathBuf,
+
+    #[serde(skip)]
+    local_version: String,
+
+    #[serde(skip)]
+    remote_repo_name: String,
+}
+
+impl Default for VersionCheckInfo {
+    fn default() -> Self {
+        Self {
+            latest_version: Default::default(),
+            contains_critical_fix: false,
+            query_time: Default::default(),
+            inform_time: DateTime::<Utc>::MIN_UTC,
+            version_check_filename: get_version_info_filename(),
+            local_version: CURRENT_VERSION.to_owned(),
+            remote_repo_name: GITHUB_REPO_NAME.to_owned(),
+        }
+    }
+}
+
+impl VersionCheckInfo {
+    fn known_new_release(&self) -> bool {
+        version_is_newer(&self.latest_version, &self.local_version)
+    }
+
+    pub fn notify_user_if_appropriate(&mut self) {
+        debug!("VersionCheckInfo:notify_user_if_appropriate: version check info = {self:?}");
+        if self.known_new_release() {
+            let mut notification_happened = false;
+
+            if self.contains_critical_fix {
+                if self.inform_age_in_seconds() >= NEW_CRITICAL_VERSION_NOTIFICATION_INTERVAL {
+                    eprintln!("\n\n**CRITICAL:**\nA new version of the Xet client tools, {}, is available at https://github.com/xetdata/xet-tools/releases and contains a critical bug fix from your current version.  It is strongly recommended to upgrade to this release immediately.\n\n",
+                              self.latest_version);
+                    notification_happened = true;
+                }
+            } else if self.inform_age_in_seconds() >= NEW_VERSION_NOTIFICATION_INTERVAL {
+                eprintln!("\nA new version of the Xet client tools, {}, is available at https://github.com/xetdata/xet-tools/releases.\n",
+                          self.latest_version);
+                notification_happened = true;
+            }
+
+            if notification_happened {
+                self.inform_time = Utc::now();
+                self.save();
+            }
+        }
+    }
+
+    pub fn query_age_in_seconds(&self) -> u64 {
+        self.query_time
+            .signed_duration_since(Utc::now())
```

Comment:
This function `signed_duration_since` should return negative value as it is `before - now`. Then taking the max between this value and 0 doesn't seem as intended.

Reply:
OH.  You are quite correct.  This could explain some intermittent issues.